File: device.py
import binascii
DEVICETYPE_METER = 0x54
DEVICETYPE_MOTION = 0x73
DEVICETYPE_CONTACT = 0x64
DEVICETYPE_CURTAIN = 0x63
DEVICETYPE_BOT = 0x48
DEVICETYPE_ENV = 0x5A


def getDevice(macaddr, rawdata):
    deviceType = int("0x"+rawdata[4:6], 16)
    if deviceType == DEVICETYPE_METER:
        return Meter(macaddr, rawdata)
    elif deviceType == DEVICETYPE_MOTION:
        return MotionSensor(macaddr, rawdata)
    elif deviceType == DEVICETYPE_CONTACT:
        return ContactSensor(macaddr, rawdata)
    elif deviceType == DEVICETYPE_CURTAIN:
        return Curtain(macaddr, rawdata)
    elif deviceType == DEVICETYPE_BOT:
        return Bot(macaddr, rawdata)
    elif deviceType == DEVICETYPE_ENV:
        return EnvSensor(macaddr, rawdata)
    else:
        return None


class Device():
    def __init__(self, macaddr, rawdata) -> None:
        self.macaddr = macaddr
        self.rawdata = rawdata

    def perse(self):
        resp = {}
        resp['sensorData'] = self.rawdata[4:]
        resp['macaddress'] = self.macaddr
        return resp


class SwitchbotDevice(Device):
    def __init__(self, macaddr, rawdata) -> None:
        super().__init__(macaddr, rawdata)

    def perse(self):
        sensorData = binascii.unhexlify(self.rawdata[4:])
        resp = super().perse()
        resp['deviceType'] = chr(getbit(sensorData[0], 6, 0))
        resp['battery'] = (getbit(sensorData[2], 6, 0))
        return resp


class MotionSensor(SwitchbotDevice):
    def __init__(self, macaddr, rawdata) -> None:
        super().__init__(macaddr, rawdata)

    def perse(self):
        sensorData = binascii.unhexlify(self.rawdata[4:])

        resp = super().perse()

        resp['scopeTested'] = bool(getbit(sensorData[1], 7))
        resp['ripState'] = getbit(sensorData[1], 6)
        resp['ledState'] = getbit(sensorData[5], 5)
        resp['iotState'] = getbit(sensorData[5], 4)
        resp['sensingDistance'] = getbit(sensorData[5], 3, 2)
        resp['lightIntensity'] = getbit(sensorData[5], 1, 0)
        resp['sinceLastTriggerRipTime'] = (sensorData[3] << 8)+sensorData[4]
        return resp


class Meter(SwitchbotDevice):
    def __init__(self, macaddr, rawdata) -> None:
        super().__init__(macaddr, rawdata)

    def perse(self):
        sensorData = binascii.unhexlify(self.rawdata[4:])

        resp = super().perse()

        resp['temperatureHighAlert'] = getbit(sensorData[3], 7)
        resp['temperatureLowAlert'] = getbit(sensorData[3], 6)
        resp['humidityLowAlert'] = getbit(sensorData[3], 5)
        resp['humidityLowAlert'] = getbit(sensorData[3], 4)

        positiveNegativeTemperatureFlag = getbit(sensorData[4], 7)
        temperature = getbit(sensorData[3], 3, 0) / \
            10 + getbit(sensorData[4], 6, 0)
        if not positiveNegativeTemperatureFlag:
            temperature = -temperature
        resp['temperature'] = temperature
        resp['humidity'] = getbit(sensorData[5], 6, 0)
        resp['temperratureScale'] = getbit(sensorData[5], 7)
        return resp


class ContactSensor(SwitchbotDevice):
    def __init__(self, macaddr, rawdata) -> None:
        super().__init__(macaddr, rawdata)

    def perse(self):
        sensorData = binascii.unhexlify(self.rawdata[4:])

        resp = super().perse()
        resp['scopeTested'] = bool((sensorData[1] & 0b10000000) >> 7)
        resp['ripState'] = (sensorData[1] & 0b01000000) >> 6
        resp['halState'] = (sensorData[3] & 0b00000110) >> 1
        resp['lightLevel'] = sensorData[3] & 0b00000001
        resp['ripUTC'] = ((sensorData[3] & 0b10000000) << 9) + \
            (sensorData[4] << 8)+sensorData[5]
        resp['halUTC'] = ((sensorData[3] & 0b01000000) << 10) + \
            (sensorData[6] << 8)+sensorData[7]
        resp['entranceCount'] = (sensorData[8] & 0b11000000) >> 6
        resp['goOutCount'] = (sensorData[8] & 0b00110000) >> 4
        resp['buttonPushCount'] = sensorData[8] & 0b00001111

        resp['scopeTested'] = bool(getbit(sensorData[1], 7))
        resp['ripState'] = getbit(sensorData[1], 6)
        resp['halState'] = getbit(sensorData[3], 2, 1)
        resp['lightLevel'] = getbit(sensorData[3], 0)
        resp['ripUTC'] = (getbit(sensorData[3], 7) << 9) + \
            (sensorData[4] << 8)+sensorData[5]
        resp['halUTC'] = (getbit(sensorData[3], 6) << 10) + \
            (sensorData[6] << 8)+sensorData[7]
        resp['entranceCount'] = getbit(sensorData[8], 7, 6)
        resp['goOutCount'] = getbit(sensorData[8], 5, 4)
        resp['buttonPushCount'] = getbit(sensorData[8], 3, 0)
        return resp


class Curtain(SwitchbotDevice):
    def __init__(self, macaddr, rawdata) -> None:
        super().__init__(macaddr, rawdata)

    def perse(self):
        sensorData = binascii.unhexlify(self.rawdata[4:])

        resp = super().perse()
        resp['wheterToAllowConnection'] = getbit(sensorData[1], 7)
        resp['calibrationSituation'] = getbit(sensorData[1], 6)
        resp['motionState'] = getbit(sensorData[3], 7)
        resp['position'] = getbit(sensorData[3], 6, 0)
        resp['lightLevel'] = getbit(sensorData[4], 7, 4)
        resp['deviceChain'] = getbit(sensorData[4], 3, 0)
        return resp


class Bot(SwitchbotDevice):
    def __init__(self, macaddr, rawdata) -> None:
        super().__init__(macaddr, rawdata)

    def perse(self):
        sensorData = binascii.unhexlify(self.rawdata[4:])

        resp = super().perse()
        resp['mode'] = getbit(sensorData[1], 7)
        resp['state'] = getbit(sensorData[1], 6)
        return resp


class EnvSensor(Device):
    def __init__(self, macaddr, rawdata) -> None:
        super().__init__(macaddr, rawdata)

    def perse(self):
        sensorData = binascii.unhexlify(self.rawdata[4:])
        resp = super().perse()
        resp['sensorData'] = self.rawdata[4:]
        resp['macaddress'] = self.macaddr
        resp['deviceType'] = chr(sensorData[0])
        resp['range'] = (sensorData[1] << 8)+sensorData[2]
        resp['tvoc'] = (sensorData[3] << 8)+sensorData[4]
        resp['co2'] = (sensorData[5] << 8)+sensorData[6]
        return resp


def getbit(char, pos_start, pos_end=-1):

    if pos_end == -1:
        pos_end = pos_start
    mask = 0
    for n in range(pos_start, pos_end-1, -1):
        mask = mask+pow(2, n)
    return ((char & mask) >> pos_end)

File: lib/device.go
package lib

import (
	"fmt"

	"github.com/go-ble/ble"
)

const (
	switchbotUUID    = "fd3d"
	switchbotBot     = 0x48
	switchbotMeter   = 0x54
	switchbotMotion  = 0x73
	switchbotContact = 0x64
	switchbotCurtain = 0x63
	m5stackEnv       = 0x5A
	m5stackInroom    = 0x58
)

type Device struct {
	Addr    string `json:"macaddress"`
	Rawdata string `json:"sensorData"`
}
type SwitchbotDevice struct {
	Device
	Battery    byte   `json:"battery"`
	DeviceType string `json:"deviceType"`
}
type Bot struct {
	SwitchbotDevice
	Mode  byte `json:"mode"`
	State byte `json:"state"`
}
type Curtain struct {
	SwitchbotDevice
	WheterToAllowConnection byte `json:"wheterToAllowConnection"`
	CalibrationSituation    byte `json:"calibrationSituation"`
	MotionState             byte `json:"motionState"`
	Position                byte `json:"position"`
	LightLevel              byte `json:"lightLevel"`
	DeviceChain             byte `json:"deviceChain"`
}
type Meter struct {
	SwitchbotDevice
	TempertureHighAlert byte    `json:"tempertureHighAlert"`
	TempertureLowAlert  byte    `json:"tempertureLowAlert"`
	HumidityHighAlert   byte    `json:"humidityHighAlert"`
	HumidityLowAlert    byte    `json:"humidityLowAlert"`
	Temperature         float32 `json:"temperature"`
	Humidity            byte    `json:"humidity"`
	TemperatureScale    byte    `json:"temperatureScale"`
}

type Motion struct {
	SwitchbotDevice
	ScopeTested             byte  `json:"scopeTested"`
	RipState                byte  `json:"ripState"`
	LedState                byte  `json:"ledState"`
	IotState                byte  `json:"iotState"`
	SensingDistance         byte  `json:"sensingDistance"`
	LightIntensity          byte  `json:"lightIntensity"`
	SinceLastTriggerRipTime int32 `json:"sinceLastTriggerRipTime"`
}

type Contact struct {
	SwitchbotDevice
	ScopeTested     byte  `json:"scopeTested"`
	RipState        byte  `json:"ripState"`
	HalState        byte  `json:"halState"`
	LightLevel      byte  `json:"lightLevel"`
	RipUTC          int32 `json:"ripUTC"`
	HalUTC          int32 `json:"halUTC"`
	EntranceCount   byte  `json:"entranceCount"`
	GooutCount      byte  `json:"gooutCount"`
	ButtonPushCount byte  `json:"buttonPushCount"`
}

// FIXME
type Sensor interface{}

func NewDevice(addr string, sd ble.ServiceData) *Device {
	return &Device{Addr: addr, Rawdata: fmt.Sprintf("%x", sd.Data)}
}
func NewSwitchbotDevice(addr string, sd ble.ServiceData) *SwitchbotDevice {
	return &SwitchbotDevice{Device: *NewDevice(addr, sd),
		DeviceType: string(sd.Data[0] & 0b01111111),
		Battery:    sd.Data[2] & 0b01111111,
	}
}
func NewBot(addr string, sd ble.ServiceData) *Bot {
	return &Bot{
		SwitchbotDevice: *NewSwitchbotDevice(addr, sd),
		Mode:            sd.Data[1] >> 7,
		State:           (sd.Data[1] & 0b01000000) >> 6,
	}
}
func NewCurtain(addr string, sd ble.ServiceData) *Curtain {
	return &Curtain{
		SwitchbotDevice:         *NewSwitchbotDevice(addr, sd),
		WheterToAllowConnection: sd.Data[1] >> 7,
		CalibrationSituation:    (sd.Data[1] & 0b01000000) >> 6,
		MotionState:             sd.Data[3] >> 7,
		Position:                sd.Data[3] & 0b01111111,
		LightLevel:              (sd.Data[4] & 0b11110000) >> 4,
		DeviceChain:             sd.Data[4] & 0b00001111,
	}
}
func NewMeter(addr string, sd ble.ServiceData) *Meter {
	return &Meter{
		SwitchbotDevice:     *NewSwitchbotDevice(addr, sd),
		TempertureHighAlert: sd.Data[3] >> 7,
		TempertureLowAlert:  (sd.Data[3] & 0b01000000) >> 6,
		HumidityHighAlert:   (sd.Data[3] & 0b000100000) >> 5,
		HumidityLowAlert:    (sd.Data[3] & 0b000010000) >> 4,
		Temperature: func() float32 {
			positiveTempFlag := sd.Data[4] >> 7
			temperature := float32(sd.Data[3]&0b00001111)/10 + float32(sd.Data[4]&0b01111111)
			if positiveTempFlag == 0 {
				temperature = -temperature
			}
			return temperature
		}(),
		Humidity:         (sd.Data[5] & 0b011111111),
		TemperatureScale: sd.Data[5] >> 7,
	}
}
func NewMotion(addr string, sd ble.ServiceData) *Motion {
	return &Motion{
		SwitchbotDevice:         *NewSwitchbotDevice(addr, sd),
		ScopeTested:             sd.Data[1] >> 7,
		RipState:                (sd.Data[1] & 0b01000000) >> 6,
		LedState:                (sd.Data[5] & 0b00100000) >> 5,
		IotState:                (sd.Data[5] & 0b00010000) >> 4,
		SensingDistance:         (sd.Data[5] & 0b00001100) >> 2,
		LightIntensity:          (sd.Data[5] & 0b00000011),
		SinceLastTriggerRipTime: (int32(sd.Data[3]) << 8) + int32(sd.Data[4]),
	}
}

func NewContact(addr string, sd ble.ServiceData) *Contact {
	return &Contact{
		SwitchbotDevice: *NewSwitchbotDevice(addr, sd),
		ScopeTested:     sd.Data[1] >> 7,
		RipState:        (sd.Data[1] & 0b01000000) >> 6,
		HalState:        (sd.Data[3] & 0b00000110) >> 1,
		LightLevel:      sd.Data[3] & 0b00000001,
		RipUTC:          int32(sd.Data[3]&0b10000000)<<9 + int32(sd.Data[4])<<8 + int32(sd.Data[5]),
		HalUTC:          int32(sd.Data[3]&0b01000000)<<10 + int32(sd.Data[6])<<8 + int32(sd.Data[7]),
		EntranceCount:   (sd.Data[8] & 0b11000000) >> 6,
		GooutCount:      (sd.Data[8] & 0b00110000) >> 4,
		ButtonPushCount: sd.Data[8] & 0b00001111,
	}
}
func GetDevice(addr string, sd ble.ServiceData) Sensor {
	if sd.UUID.String() == switchbotUUID {
		switch sd.Data[0] {
		case switchbotBot:
			return NewBot(addr, sd)
		case switchbotCurtain:
			return NewCurtain(addr, sd)
		case switchbotMeter:
			return NewMeter(addr, sd)
		case switchbotMotion:
			return NewMotion(addr, sd)
		case switchbotContact:
			return NewContact(addr, sd)
		default:
			return nil
		}
	}
	return nil
}
